pkg/cassandra/datacenter.go

```diff
@@ -0,0 +1,141 @@
+package cassandra
+
+import (
+	cassdcapi "github.com/k8ssandra/cass-operator/operator/pkg/apis/cassandra/v1beta1"
+	api "github.com/k8ssandra/k8ssandra-operator/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// SystemReplication represents the replication factor of the system_auth, system_traces,
+// and system_distributed keyspsces. This is applied to each datacenter. The replication
+// should be configured per DC, but that is currently not supported. See
+// https://github.com/k8ssandra/management-api-for-apache-cassandra/issues/124 and
+// https://github.com/k8ssandra/k8ssandra-operator/issues/91 for details.
+// Note that when we can configure the replication per DC, this can be changed to a
+// map[string]int.
+type SystemReplication struct {
+	Datacenters       []string
+	ReplicationFactor int
+}
+
+// DatacenterConfig provides the configuration to be applied to the CassandraDatacenter.
+// A DatacenterConfig is essentially a coalescence of an api.CassandraClusterTemplate and
+// an api.CassandraDatacenterTemplate. There are global, cluster-wide settings that need
+// to be specified at the DC-level. Using a DatacenterConfig allows to keep the api types
+// clean such that cluster-level settings won't leak into the dc-level settings.
+type DatacenterConfig struct {
+	Meta              api.EmbeddedObjectMeta
+	Cluster           string
+	ServerImage       string
+	ServerVersion     string
+	Size              int32
+	Resources         *corev1.ResourceRequirements
+	SystemReplication SystemReplication
+	StorageConfig     *cassdcapi.StorageConfig
+	Racks             []cassdcapi.Rack
+	CassandraConfig   *api.CassandraConfig
+	AdditionalSeeds   []string
+}
+
+func NewDatacenter(klusterKey types.NamespacedName, template *DatacenterConfig) (*cassdcapi.CassandraDatacenter, error) {
+	namespace := template.Meta.Namespace
+	if len(namespace) == 0 {
```

Comment:
if namespace == "" ?

Reply:
What is the question? Are you asking why I am checking length vs `namespace == ""`? If so both are idiomatic approaches.

Reply:
With the exception that this one doesn't catch diff between []string and string